### PR TITLE
NMS-10598: Add scope for node/interface/service details

### PR DESCRIPTION
**Pull request by fooker**
* JIRA: http://issues.opennms.org/browse/NMS-10598

### CODE REVIEW

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java [hunk 1]
```diff
@@ -89,14 +106,22 @@ public Scope getScopeForInterface(final Integer nodeId, final String ipAddress)
...
             }
 
-            return new SimpleScope(transform(ipInterface.getMetaData()));
+            return new FallbackScope(transform(ipInterface.getMetaData()),
+                    new ObjectScope<>(ipInterface)
+                            .map("interface", "hostname", (i) -> Optional.of(i.getIpHostName()))
+                            .map("interface", "address", (i) -> Optional.of(i.getIpAddress()).map(InetAddressUtils::toIpAddrString))
+                            .map("interface", "netmask", (i) -> Optional.of(i.getNetMask()).map(InetAddressUtils::toIpAddrString))
+                            .map("interface", "if-index", (i) -> Optional.ofNullable(i.getIfIndex()).map(Object::toString))
+                            .map("interface", "if-alias", (i) -> Optional.of(i.getSnmpInterface().getIfAlias()))
```
Comment (christianpape) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java]: Check for Null here. A SnmpInterface will not exist in any case.

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java [hunk 2]
```diff
@@ -109,15 +134,19 @@ public Scope getScopeForService(final Integer nodeId, final InetAddress ipAddres
                 return EmptyScope.EMPTY;
             }
 
-            return new SimpleScope(transform(monitoredService.getMetaData()));
+            return new FallbackScope(transform(monitoredService.getMetaData()),
+                    new ObjectScope<>(monitoredService)
+                            .map("service", "name", (s) -> Optional.of(s.getServiceName()))
+                            .map("service", "type", (s) -> Optional.of(s.getServiceType().getName()))
```
Comment (christianpape) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java]: `getServiceName() == getServiceType().getName()`
Does this make sense?

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java [hunk 3]
```diff
@@ -109,15 +134,19 @@ public Scope getScopeForService(final Integer nodeId, final InetAddress ipAddres
...
         });
 
         return metaDataScope;
     }
 
-    private static Map<ContextKey, String> transform(Collection<OnmsMetaData> metaData) {
-        return metaData.stream().collect(Collectors.toMap(
-                        e -> new ContextKey(e.getContext(), e.getKey()),
-                        e -> e.getValue()));
+    private static MapScope transform(Collection<OnmsMetaData> metaData) {
```
Comment (christianpape) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java]: final

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 4]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+In relation to nodes, interfaces and services, there are special contexts available which expose details about those objects.
+
+The `node` context provides details about the node currently processed.
+The following keys are available under this context:
+
+[options="header, autowidth"]
+|===
+| Context:Key            | Description |
+| `node:label`           | The label of the node
+| `node:foreign-source`  | The nodes foreign source name
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: node's

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 5]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+| `node:foreign-id`      | The nodes foreigne ID
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: same here

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 6]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+| `node:netbios-domain`  | The NetBIOS domain as provided by SNMP
+| `node:netbios-name`    | The NetBIOS name as provided by SNMP
+| `node:os`              | The nodes operating system
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: again

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 7]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+| `node:sys-name`        | The system name of the node
+| `node:sys-location`    | The system localtion of the node
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: location

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 8]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+| `node:sys-contact`     | The system contact specified for the node
+| `node:sys-description` | The system description of the node
+| `node:location`        | The nodes monitoring location name
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: 's

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 9]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+| `node:area`            | The nodes monitoring location area
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: 's

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 10]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+|===
+
+The `interface` context provides details about the interface currently processed.
+The following keys are available unter this context:
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: unter?

File: opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc [hunk 11]
```diff
@@ -12,6 +12,54 @@ These pages show the meta-data as it is currently assigned in the database.
...
+| `interface:address        | The IP address if the interface
+| `interface:netmask        | The netmask of the interface
+| `interface:if-index       | The SNMP interface index
+| `interface:if-alias       | The SNMP interface alias
+| `interface:if-description | The SNMP interface description
+| `interface:phy-addr       | The physical address of the interface
+|===
+
+The `service` context provides details about the service currently processed.
+The following keys are available unter this context:
```
Comment (christianpape) [inline opennms-doc/guide-admin/src/asciidoc/text/meta-data.adoc]: unter?

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java [hunk 12]
```diff
@@ -31,19 +31,22 @@
...
 import org.opennms.netmgt.model.OnmsMonitoredService;
 import org.opennms.netmgt.model.OnmsNode;
-import org.opennms.netmgt.model.OnmsMetaData;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.transaction.support.TransactionTemplate;
 
 import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
 
 public class EntityScopeProviderImpl implements EntityScopeProvider {
```
Comment (christianpape) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java]: Please add tests :-)

Comment (fooker) [reply core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java]: Added tests...

**Pull request merged**